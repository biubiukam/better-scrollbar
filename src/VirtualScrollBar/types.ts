import { HTMLProps, PropsWithChildren, ReactElement } from "react"

/** 滚动状态 */
export interface ScrollState {
	/** X轴滚动偏移 */
	x: number
	/** Y轴滚动偏移 */
	y: number
	/** 可滚动宽度 */
	scrollWidth: number
	/** 可滚动高度 */
	scrollHeight: number
	/** 滚动视区宽度 */
	clientWidth: number
	/** 滚动视区高度 */
	clientHeight: number
	/** 是否正在滚动 */
	isScrolling: boolean
}

/** 滚动偏移 */
export interface ScrollOffset {
	/** X轴上的偏移 */
	x: number
	/** Y轴上的偏移 */
	y: number
}

/** 渲染元素 */
export type RenderElement<Props> = (props?: PropsWithChildren<Props>) => ReactElement

// (props?: PropsWithChildren<Props>) => ReactElement | ForwardRefExoticComponent<PropsWithoutRef<Instance> & RefAttributes<PropsWithChildren<Props>>>

/** 组件Props */
export interface VirtualScrollBarProps {
	/** 开始滚动回调 */
	onScrollStart?: () => void
	/** 结束滚动回调 */
	onScrollEnd?: () => void
	/**
	 * @description 滚动回调
	 * @param {ScrollState} scrollState 滚动状态
	 */
	onScroll?: (scrollState: ScrollState) => void
	/** 是否需要虚拟滚动 */
	isVirtual?: boolean
	/** 外层容器样式 */
	className?: string
	/** 单条数据默认高度 */
	itemHeight?: number
	/** 样式前缀 */
	prefixCls?: string
	/** 滚动容器宽度 */
	width?: number
	/** 滚动容器高度 */
	height?: number
	/** 滚动条粗细 */
	scrollBarSize?: number
	/** 滚动条是否隐藏 */
	scrollBarHidden?: boolean
	/** 滚动条隐藏延时 */
	scrollBarAutoHideTimeout?: number
	/**
	 * @description 绘制滚动区域
	 * @param {(HTMLAttributes<HTMLDivElement>) => React.ReactElement} props
	 */
	renderView?: RenderElement<HTMLProps<HTMLDivElement>>
}

export interface VirtualScrollBarRef {
	/** 滚动到指定位置 */
	scrollTo: (offset: ScrollOffset) => void
	/** 获取当前的滚动数据 */
	getScrollState: () => ScrollState
	/** 滚动、视区中的高宽变化回调 */
	resizeObserver: (callback: (resizeState: Pick<ScrollState, "scrollWidth" | "scrollHeight" | "clientWidth" | "clientHeight">) => void) => void
}
