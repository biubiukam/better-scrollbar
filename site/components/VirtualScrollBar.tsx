import React, { PropsWithChildren, useCallback, useLayoutEffect, useRef, useState } from "react"
import ScrollBars, { type ScrollbarProps } from "../../src"

const itemHeight = 40

function VirtualScrollBar({className, children, ...props}: PropsWithChildren<ScrollbarProps>) {
	
	const domRef = useRef<HTMLDivElement>({} as HTMLDivElement)
	
	const [itemIndex, setIndex] = useState<[number, number]>([0, 0])
	
	const render = useCallback((scrollTop = 0) => {
		// @ts-ignore
		const {height = 0} = domRef.current?.getBoundingClientRect?.() || {}
		
		const visibleItemCount = Math.ceil(height / itemHeight)
		const startIndex = Math.floor(scrollTop / itemHeight)
		const endIndex = startIndex + visibleItemCount
		
		setIndex([startIndex, endIndex])
	}, [])
	
	const onScrollFrame = useCallback((position) => {
		render(position.scrollTop)
	}, [])
	
	useLayoutEffect(() => {
		render()
	}, [children])
	
	return (
		<div className={ className } ref={ domRef } style={{height: "100%"}}>
			<ScrollBars
				style={ {height: "100%"} }
				onScrollFrame={ onScrollFrame }
				{ ...props }
			>
				<div style={ {height: React.Children.count(children) * itemHeight} }/>
				{
					React.Children.map(children as React.ReactChildren | React.ReactChildren[], (child: any, index) => {
						if (index >= itemIndex[0] && index <= itemIndex[1]) {
							return React.cloneElement(child, {
								style: {
									...(child?.props?.style || {}),
									transform: `translateY(${ index * itemHeight }px)`
								}
							})
						}
						return null
					})
				}
			</ScrollBars>
		</div>
	)
}

export default VirtualScrollBar
