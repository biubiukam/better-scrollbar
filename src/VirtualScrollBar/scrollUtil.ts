const MIN_SIZE = 25;
/**
 * @description 获取滚动条的大小
 * @param {number} containerSize 滚动视区高度
 * @param {number} scrollRange 实际滚动高度
 */
export function getSpinSize(containerSize: number = 0, scrollRange: number = 0) {
	let baseSize = (containerSize / scrollRange) * 100;
	if (isNaN(baseSize)) {
		baseSize = 0;
	}
	baseSize = Math.max(baseSize, MIN_SIZE);
	baseSize = Math.min(baseSize, containerSize / 2);
	return Math.floor(baseSize);
}

