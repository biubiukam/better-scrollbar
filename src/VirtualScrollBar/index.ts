import VirtualScrollBar from "./VirtualScrollBar"

export type { ScrollState } from "./types"
export type { ScrollBarProps } from "./VirtualScrollBar"
export default VirtualScrollBar
