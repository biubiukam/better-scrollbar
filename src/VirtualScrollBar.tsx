import React, {
	useCallback,
	useEffect,
	useMemo,
	useRef,
	Children, forwardRef, useImperativeHandle, cloneElement
} from "react"
import type { PropsWithChildren } from "react"
import raf from "./raf"
import { useImmer } from "use-immer"
import { Item } from "./components/Item"
import useHeights from "./hooks/useHeights"
import VerticalScrollBar from "./components/VerticalScrollBar"
import { getSpinSize } from "./scrollUtil"
import { ScrollOffset, ScrollState, VirtualScrollBarProps, VirtualScrollBarRef, ScrollBarRef } from "./types"
import useResizeObserver from "./hooks/useResizeObserver"
import clsx from "clsx"
import {
	renderViewDefault,
	renderTrackVerticalDefault,
	renderThumbVerticalDefault
} from "./defaultRenderElements"

const VirtualScrollBar = forwardRef<VirtualScrollBarRef, PropsWithChildren<VirtualScrollBarProps>>((props, ref) => {
	const {
		onScrollStart,
		onScrollEnd,
		onScroll,
		children,
		width,
		height,
		className,
		prefixCls = "scroll-bar",
		itemHeight = 20,
		scrollBarSize = 6,
		scrollBarHidden = false,
		scrollBarAutoHideTimeout = 1000,
		renderView = renderViewDefault,
		renderTrackVertical = renderTrackVerticalDefault,
		renderThumbVertical = renderThumbVerticalDefault,
		// renderThumbHorizontal = renderThumbHorizontalDefault,
		// renderTrackHorizontal = renderTrackHorizontalDefault
	} = props
	
	const childNodes = useMemo(() => {
		return (typeof children === "function" ? [children] : Children.toArray(children)) as Array<React.ReactElement>
	}, [children])
	
	// 可见视图区域
	const viewContainerRef = useRef<HTMLDivElement>({} as HTMLDivElement)
	// 滚动区域
	const scrollContainerRef = useRef<HTMLDivElement>({} as HTMLDivElement)
	// 滚动条
	const verticalScrollBarInstance = useRef<ScrollBarRef>({} as ScrollBarRef)
	const {setInstanceRef, collectHeight, heights, updatedMark} = useHeights()
	
	const [scrollState, setScrollState] = useImmer<ScrollState>({
		x: 0,
		y: 0,
		scrollHeight: 0,
		scrollWidth: 0,
		clientHeight: 0,
		clientWidth: 0,
		isScrolling: false
	})
	
	useResizeObserver(viewContainerRef, (newSize) => {
		setScrollState((preScrollState) => {
			preScrollState.clientHeight = newSize.height
			preScrollState.clientWidth = newSize.width
		})
	})
	
	const {scrollHeight, start, end, offset: fillerOffset} = useMemo(() => {
		let itemTop = 0
		let startIndex: number | undefined
		let startOffset: number | undefined
		let endIndex: number | undefined
		
		for (let i = 0, len = childNodes.length; i < len; i++) {
			const key = childNodes[i]?.key as React.Key
			
			const cacheHeight = heights.get(key)
			const currentItemBottom = itemTop + (cacheHeight === undefined ? itemHeight : cacheHeight)
			
			// 选中范围中的顶部项目
			if (currentItemBottom >= scrollState.y && startIndex === undefined) {
				startIndex = i
				startOffset = itemTop
			}
			
			// 检查范围内的项目底部。我们将渲染额外的一个项目以供运动使用
			if (currentItemBottom > scrollState.y + scrollState.clientHeight && endIndex === undefined) {
				endIndex = i
			}
			
			itemTop = currentItemBottom
		}
		// 当滚动顶部在末尾，但数据被剪切到小计数时，将达到此值
		if (startIndex === undefined) {
			startIndex = 0
			startOffset = 0
			
			endIndex = Math.ceil(scrollState.clientHeight / itemHeight)
		}
		if (endIndex === undefined) {
			endIndex = childNodes.length - 1
		}
		
		// 提供缓存以改善滚动体验
		endIndex = Math.min(endIndex + 1, childNodes.length - 1)
		return {
			scrollHeight: itemTop,
			start: startIndex,
			end: endIndex,
			offset: startOffset,
		}
	}, [itemHeight, childNodes, scrollState.y, updatedMark, scrollState.clientHeight])
	
	useEffect(() => {
		setScrollState((preScrollState) => {
			preScrollState.scrollHeight = scrollHeight
		})
	}, [scrollHeight])
	
	const maxScrollHeight = scrollHeight - scrollState.clientHeight
	const maxScrollHeightRef = useRef(maxScrollHeight)
	maxScrollHeightRef.current = maxScrollHeight
	
	const keepInRange = useCallback((newScrollTop: number) => {
		let newTop = newScrollTop
		if (!Number.isNaN(maxScrollHeightRef.current)) {
			newTop = Math.min(newTop, maxScrollHeightRef.current)
		}
		newTop = Math.max(newTop, 0)
		return newTop
	}, [])
	
	const detectScrollingInterval = useRef<ReturnType<typeof setTimeout>>()
	
	const onUpdateScrollState = useCallback((scrollTop: number | ((preScrollTop: number) => number)) => {
		setScrollState((preScrollState) => {
			if (typeof scrollTop === "function") {
				preScrollState.y = scrollTop(preScrollState.y)
			} else {
				preScrollState.y = scrollTop
			}
			preScrollState.isScrolling = true
			viewContainerRef.current.scrollTop = preScrollState.y
		})
		
		delayScrollStateChange()
	}, [])
	
	useEffect(() => {
		onScroll?.(scrollState)
	}, [scrollState])
	
	/**
	 * @description 延迟"是否滚动"的滚动状态变更
	 */
	const delayScrollStateChange = useCallback(() => {
		clearTimeout(detectScrollingInterval.current)
		
		detectScrollingInterval.current = setTimeout(() => {
			setScrollState((preScrollState) => {
				preScrollState.isScrolling = false
			})
		}, 200)
	}, [])
	
	useEffect(() => {
		if (scrollState.isScrolling) {
			onScrollStart?.()
		} else {
			onScrollEnd?.()
		}
	}, [scrollState.isScrolling])
	
	const wheelingRaf = useRef<number>(-1)
	
	useEffect(() => {
		const onScroll = function (event: WheelEvent): void {
			event?.preventDefault()
			
			const {deltaX, deltaY} = event
			
			const StepY = 360
			const StepX = 360
			
			// 滚动方向重置以及滚动大小控制  按住 shift 时，调换滚轮方向
			const scrollTop = event.shiftKey
				? Math.max(Math.min(deltaX, StepX), -StepX)
				: Math.max(Math.min(deltaY, StepY), -StepY)
			
			wheelingRaf.current = raf(() => {
				raf.cancel(wheelingRaf.current)
				collectHeight()
				
				onUpdateScrollState((preScrollStateY) => {
					return keepInRange(Math.max(preScrollStateY + scrollTop, 0))
				})
			})
		}
		const scrollContainer = scrollContainerRef.current;
		scrollContainer?.addEventListener("wheel", onScroll)
		return () => {
			scrollContainer?.removeEventListener("wheel", onScroll)
			raf.cancel(wheelingRaf.current)
		}
	}, [])
	
	const listChildren = useMemo(() => {
		return childNodes.slice(start, end + 1).map((node) => {
			return (
				<Item key={ node?.key } setRef={ (ele) => setInstanceRef(node, ele) }>
					{ node }
				</Item>
			)
		})
	}, [childNodes, start, end, setInstanceRef])
	
	const delayHideScrollBar = useCallback(() => {
		verticalScrollBarInstance.current?.delayHiddenScrollBar()
	}, [])
	
	// 当数据大小减小时。它可能会触发本地滚动事件以适应滚动位置
	const onFallbackScroll = useCallback((event: React.UIEvent<HTMLDivElement>) => {
		const {scrollTop: newScrollTop} = event.currentTarget
		if (newScrollTop !== scrollState.y) {
			onUpdateScrollState(newScrollTop)
		}
	}, [scrollState])
	
	useImperativeHandle(ref, (): VirtualScrollBarRef => {
		return {
			scrollTo(offset: ScrollOffset) {
				onUpdateScrollState(offset.y)
			},
			getScrollState() {
				return scrollState
			},
			resizeObserver(callback) {
				callback({
					clientWidth: scrollState.clientWidth,
					clientHeight: scrollState.clientHeight,
					scrollWidth: scrollState.scrollWidth,
					scrollHeight: scrollState.scrollHeight,
				})
			}
		}
	})
	
	return (
		<div style={ {width, height} } className={ clsx(className, `${ prefixCls }-outer-container`) }>
			<div
				ref={ viewContainerRef }
				style={ {width, height} }
				className={ clsx(`${ prefixCls }-inner-container`) }
				onMouseEnter={ delayHideScrollBar }
				onScroll={ onFallbackScroll }
			>
				<div
					ref={ scrollContainerRef }
					style={ {height: scrollHeight} }
					className={ clsx(`${ prefixCls }-container`) }
					onScroll={ event => event.preventDefault() }
				>
					{
						cloneElement(
							renderView({
								className: clsx(`${ prefixCls }-wrapper`),
								style: {transform: `translateY(${ fillerOffset }px)`}
							}),
							{},
							listChildren
						)
					}
				</div>
			</div>
			<VerticalScrollBar
				prefixCls={ prefixCls }
				ref={ verticalScrollBarInstance }
				hidden={ scrollBarHidden }
				thumbSize={ {
					height: getSpinSize(scrollState.clientHeight, scrollHeight),
					width: scrollBarSize
				} }
				renderTrack={ renderTrackVertical }
				renderThumb={ renderThumbVertical }
				autoHideTimeout={ scrollBarAutoHideTimeout }
				scrollState={ scrollState }
				scrollRange={ scrollHeight }
				containerSize={ scrollState.clientHeight }
				onScroll={ onUpdateScrollState }
			/>
		</div>
	)
})

export default VirtualScrollBar
