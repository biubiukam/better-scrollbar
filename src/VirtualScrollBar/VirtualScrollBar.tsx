import React, {
	PropsWithChildren,
	useCallback,
	useEffect,
	useMemo,
	useRef,
	Children,
	useState,
} from "react"
import { flushSync } from "react-dom"
import raf from "./raf"
import { useImmer } from "use-immer"
import { Item } from "./Item"
import useHeights from "./hooks/useHeights"
import VerticalScrollBar, { VerticalScrollBarRef } from "./VerticalScrollBar"
import { getSpinSize } from "./scrollUtil"
import { ScrollState, Size } from "./types"
import useResizeObserver from "./hooks/useResizeObserver"
import clsx from "clsx"
import "./init.less"

/**
 * props 汇总
 * @param {} onScrollStart 开始滚动回调
 * @param {} onScrollEnd 结束滚动回调
 * @param {} onScroll 滚动更新回调（返回当前滚动的x、y，意味着就是top、left）
 * @param {number} thumbSize 滚动条粗细（配置看看能不能聚合）
 * @param {number} thumbAutoHideTimeout 滚动条隐藏延时
 * @param {number} thumbMinSize 滚动条最小面积
 * @param {boolean} isVirtual 是否需要虚拟滚动
 * @constructor
 */
export interface ScrollBarProps {
	/** 开始滚动回调 */
	onScrollStart?: () => void
	/** 结束滚动回调 */
	onScrollEnd?: () => void
	/** 滚动更新回调 */
	onScroll?: (scrollState: ScrollState) => void
	/** 滚动条粗细 */
	thumbSize?: number
	/** 滚动条隐藏延时 */
	thumbAutoHideTimeout?: number
	/** 滚动条最小长度 */
	thumbMinSize?: number
	/** 是否需要虚拟滚动 */
	isVirtual?: boolean
	/** 外层容器样式 */
	className?: string
	/** 样式前缀 */
	prefixCls?: string
	/** 滚动容器宽度 */
	width?: number
	/** 滚动容器高度 */
	height?: number
	/** 当前可查看的数据 */
	onVisibleChange?: () => void
}

function VirtualScrollBar(props: PropsWithChildren<ScrollBarProps>) {
	const {
		onScrollStart,
		onScrollEnd,
		onScroll,
		children,
		width,
		height,
		className,
		prefixCls = "scroll-bar"
	} = props
	
	const childNodes = useMemo(() => {
		return (typeof children === "function" ? [children] : Children.toArray(children)) as Array<React.ReactElement>
	}, [children])
	
	// 滚动视区高宽
	const [size, setSize] = useState<Size>({width: 0, height: 0})
	// 可见视图区域
	const viewContainerRef = useRef<HTMLDivElement>({} as HTMLDivElement)
	// 滚动区域
	const scrollContainerRef = useRef<HTMLDivElement>({} as HTMLDivElement)
	// 滚动条
	const verticalScrollBarInstance = useRef<VerticalScrollBarRef>({} as VerticalScrollBarRef)
	// const horizontalScrollBarInstance = useRef<HTMLDivElement>({} as HTMLDivElement)
	const {setInstanceRef, collectHeight, heights, updatedMark} = useHeights()
	
	const [scrollState, setScrollState] = useImmer<ScrollState>({
		x: 0,
		y: 0,
		isScrolling: false
	})
	
	useResizeObserver(viewContainerRef, (newSize) => {
		setSize(newSize)
	})
	
	const {scrollHeight, start, end, offset: fillerOffset} = useMemo(() => {
		let itemTop = 0
		let startIndex: number | undefined
		let startOffset: number | undefined
		let endIndex: number | undefined
		
		for (let i = 0, len = childNodes.length; i < len; i++) {
			const key = childNodes[i]?.key as React.Key
			
			const cacheHeight = heights[key]
			const currentItemBottom = itemTop + (cacheHeight === undefined ? 10 : cacheHeight)
			
			// Check item top in the range
			if (currentItemBottom >= scrollState.y && startIndex === undefined) {
				startIndex = i
				startOffset = itemTop
			}
			
			// Check item bottom in the range. We will render additional one item for motion usage
			if (currentItemBottom > scrollState.y + size.height && endIndex === undefined) {
				endIndex = i
			}
			
			itemTop = currentItemBottom
		}
		// When scrollTop at the end but data cut to small count will reach this
		if (startIndex === undefined) {
			startIndex = 0
			startOffset = 0
			
			endIndex = Math.ceil(size.height / 10)
		}
		if (endIndex === undefined) {
			endIndex = childNodes.length - 1
		}
		
		// Give cache to improve scroll experience
		endIndex = Math.min(endIndex + 1, childNodes.length - 1)
		return {
			scrollHeight: itemTop,
			start: startIndex,
			end: endIndex,
			offset: startOffset,
		}
	}, [childNodes, scrollState.y, updatedMark, size.height])
	
	const maxScrollHeight = scrollHeight - size.height
	const maxScrollHeightRef = useRef(maxScrollHeight)
	maxScrollHeightRef.current = maxScrollHeight
	
	function keepInRange(newScrollTop: number) {
		let newTop = newScrollTop
		if (!Number.isNaN(maxScrollHeightRef.current)) {
			newTop = Math.min(newTop, maxScrollHeightRef.current)
		}
		newTop = Math.max(newTop, 0)
		return newTop
	}
	
	const detectScrollingInterval = useRef<ReturnType<typeof setTimeout>>()
	
	const onUpdateScrollState = useCallback((scrollTop: number | ((preScrollTop: number) => number)) => {
		setScrollState((preScrollState) => {
			if (typeof scrollTop === "function") {
				preScrollState.y = scrollTop(preScrollState.y)
			} else {
				preScrollState.y = scrollTop
			}
			preScrollState.isScrolling = true
			viewContainerRef.current.scrollTop = preScrollState.y
		})
		
		// 同步滚动状态
		flushSync(() => {
			onScroll?.(scrollState)
		})
		delayScrollStateChange()
	}, [])
	
	/**
	 * @description 延迟"是否滚动"的滚动状态变更
	 */
	const delayScrollStateChange = useCallback(() => {
		clearTimeout(detectScrollingInterval.current)
		
		detectScrollingInterval.current = setTimeout(() => {
			setScrollState((preScrollState) => {
				preScrollState.isScrolling = false
			})
		}, 200)
	}, [])
	
	useEffect(() => {
		if (scrollState.isScrolling) {
			onScrollStart?.()
		} else {
			onScrollEnd?.()
		}
	}, [scrollState.isScrolling])
	
	const wheelingRaf = useRef<number>(-1)
	
	useEffect(() => {
		const onScroll = function (event: WheelEvent): void {
			event?.preventDefault()
			
			const {deltaX, deltaY} = event
			
			const StepY = 360
			const StepX = 360
			
			// 滚动方向重置以及滚动大小控制  按住 shift 时，调换滚轮方向
			const scrollTop = event.shiftKey
				? Math.max(Math.min(deltaX, StepX), -StepX)
				: Math.max(Math.min(deltaY, StepY), -StepY)
			
			wheelingRaf.current = raf(() => {
				raf.cancel(wheelingRaf.current)
				collectHeight()
				
				onUpdateScrollState((preScrollStateY) => {
					return keepInRange(Math.max(preScrollStateY + scrollTop, 0))
				})
			})
		}
		
		scrollContainerRef.current?.addEventListener("wheel", onScroll)
		return () => {
			raf.cancel(wheelingRaf.current)
			scrollContainerRef.current?.removeEventListener("wheel", onScroll)
		}
	}, [])
	
	const listChildren = useMemo(() => {
		return childNodes.slice(start, end + 1).map((node) => {
			return (
				<Item key={ node?.key } setRef={ (ele) => setInstanceRef(node, ele) }>
					{ node }
				</Item>
			)
		})
	}, [childNodes, start, end, setInstanceRef])
	
	const delayHideScrollBar = useCallback(() => {
		verticalScrollBarInstance.current?.delayHiddenScrollBar()
	}, [])
	
	// When data size reduce. It may trigger native scroll event back to fit scroll position
	const onFallbackScroll = useCallback((event: React.UIEvent<HTMLDivElement>) => {
		const {scrollTop: newScrollTop} = event.currentTarget
		if (newScrollTop !== scrollState.y) {
			onUpdateScrollState(newScrollTop)
		}
	}, [scrollState])
	
	return (
		<div style={ {width, height} } className={ clsx(className, `${ prefixCls }-outer-container`) }>
			<div
				ref={ viewContainerRef }
				style={ {width, height} }
				className={ clsx(`${ prefixCls }-inner-container`) }
				onMouseEnter={ delayHideScrollBar }
				onScroll={ onFallbackScroll }
			>
				<div
					ref={ scrollContainerRef }
					style={ {height: scrollHeight} }
					className={ clsx(`${ prefixCls }-container`) }
					onScroll={ event => event.preventDefault() }
				>
					<div
						className={ clsx(`${ prefixCls }-wrapper`) }
						style={ {transform: `translateY(${ fillerOffset }px)`} }
					>
						{ listChildren }
					</div>
				</div>
			</div>
			<VerticalScrollBar
				prefixCls={ prefixCls }
				ref={ verticalScrollBarInstance }
				scrollState={ scrollState }
				scrollRange={ scrollHeight }
				containerSize={ size.height }
				spinSize={ getSpinSize(size.height, scrollHeight) }
				onScroll={ onUpdateScrollState }
			/>
		</div>
	)
}

export default VirtualScrollBar
