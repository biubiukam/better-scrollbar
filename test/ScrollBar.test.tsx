import { render, act } from "@testing-library/react"
import VirtualScrollBar, { VirtualScrollBarRef } from "../src"
import React, { createRef } from "react"
import { expect, describe, it, vi } from "vitest"
import "../src/styles/index.less"

const mockData = Array.from({ length: 20 }, (_, i) => i)

describe("VirtualScrollBar", () => {
	it("matches snapshot", () => {
		const { container } = render(
			<VirtualScrollBar width={600} height={600}>
				{mockData.map((i) => (
					<div key={i} style={{ height: 50 }}>
						{i}
					</div>
				))}
			</VirtualScrollBar>
		)
		expect(container.firstChild).toMatchSnapshot()
	})

	it("renders without crashing", () => {
		const { container } = render(<VirtualScrollBar />)
		expect(container.querySelector(".scroll-bar-wrapper")).not.toBeNull()
	})

	it("calls onScrollStart and onScrollEnd when scrolling starts and ends", () => {
		vi.useFakeTimers()
		const handleScrollStart = vi.fn()
		const preventDefault = vi.fn()
		const { container } = render(
			<VirtualScrollBar width={100} height={100} onScrollStart={handleScrollStart}>
				{mockData.map((i) => (
					<div key={i} style={{ height: 50 }}>
						{i}
					</div>
				))}
			</VirtualScrollBar>
		)
		const scrollContainer = container.querySelector(".scroll-bar-container") as Element
		const wheelEvent = new WheelEvent("wheel", { deltaY: 100 })
		wheelEvent.preventDefault = preventDefault
		act(() => {
			scrollContainer.dispatchEvent(wheelEvent)
		})
		expect(preventDefault).toHaveBeenCalledTimes(1)
		vi.useRealTimers()
	})

	it("renders without children", () => {
		const { container } = render(<VirtualScrollBar />)
		expect(container.firstChild).not.toBeNull()
	})

	it("renders with multiple children", () => {
		const { container } = render(
			<VirtualScrollBar>
				<div>Child 1</div>
				<div>Child 2</div>
				<div>Child 3</div>
			</VirtualScrollBar>
		)
		expect(container.firstChild).not.toBeNull()
	})

	it("updates scroll position when scrolling", () => {
		vi.useFakeTimers()
		const ref = createRef<VirtualScrollBarRef>()
		const { container } = render(
			<VirtualScrollBar height={50} ref={ref}>
				{mockData.map((i) => (
					<div key={i} style={{ height: 50 }}>
						{i}
					</div>
				))}
			</VirtualScrollBar>
		)
		const scrollContainer = container.querySelector(".scroll-bar-container") as Element
		const scrollY = 100
		const wheelEvent = new WheelEvent("wheel", { deltaY: scrollY })
		act(() => {
			scrollContainer.dispatchEvent(wheelEvent)
			vi.runAllTimers()
		})
		const { y } = ref.current?.getScrollState() || {}
		expect(y).toBe(scrollY)
		vi.useRealTimers()
	})

	it("updates scroll position when window resizes", () => {
		vi.useFakeTimers()
		const ref = createRef<VirtualScrollBarRef>()
		const { container } = render(
			<VirtualScrollBar ref={ref}>
				{mockData.map((i) => (
					<div key={i} style={{ height: 50 }}>
						{i}
					</div>
				))}
			</VirtualScrollBar>
		)
		const scrollContainer = container.querySelector(".scroll-bar-container") as Element
		const initialScrollTop = scrollContainer.scrollHeight
		act(() => {
			window.dispatchEvent(new Event("resize"))
			vi.runAllTimers()
		})
		const { scrollHeight } = ref.current?.getScrollState() || {}
		expect(scrollHeight).not.toBe(initialScrollTop)
		vi.useRealTimers()
	})

	it("use ref to scroll to a specific position", () => {
		vi.useFakeTimers()
		const ref = createRef<VirtualScrollBarRef>()
		render(
			<VirtualScrollBar ref={ref}>
				{mockData.map((i) => (
					<div key={i} style={{ height: 50 }}>
						{i}
					</div>
				))}
			</VirtualScrollBar>
		)
		const scrollY = 200
		act(() => {
			ref.current?.scrollTo({ x: 0, y: scrollY })
			vi.runAllTimers()
		})
		const { y } = ref.current?.getScrollState() || {}
		expect(y).toBe(scrollY)
		vi.useRealTimers()
	})
})
