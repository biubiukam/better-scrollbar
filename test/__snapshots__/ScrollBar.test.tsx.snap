// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VirtualScrollBar > matches snapshot 1`] = `
<div
  class="scroll-bar-outer-container"
  style="width: 600px; height: 600px;"
>
  <div
    class="scroll-bar-inner-container"
    style="width: 600px; height: 600px;"
  >
    <div
      class="scroll-bar-container"
      style="height: 400px;"
    >
      <div
        class="scroll-bar-wrapper"
        style="transform: translateY(0px);"
      >
        <div
          style="height: 50px;"
        >
          0
        </div>
        <div
          style="height: 50px;"
        >
          1
        </div>
      </div>
    </div>
  </div>
  <div
    class="scroll-bar-vertical-track"
    style="width: 6px; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
  >
    <div
      class="scroll-bar-vertical-thumb"
      style="transform: translateY(0px); height: 0px; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
    />
  </div>
</div>
`;

exports[`VirtualScrollBar matches snapshot 1`] = `
<div
  class="scroll-bar-outer-container"
  style="width: 600px; height: 600px;"
>
  <div
    class="scroll-bar-inner-container"
    style="width: 600px; height: 600px;"
  >
    <div
      class="scroll-bar-container"
      style="height: 400px;"
    >
      <div
        class="scroll-bar-wrapper"
        style="transform: translateY(0px);"
      >
        <div
          style="height: 50px;"
        >
          0
        </div>
        <div
          style="height: 50px;"
        >
          1
        </div>
      </div>
    </div>
  </div>
  <div
    class="scroll-bar-vertical-track"
    style="width: 6px; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
  >
    <div
      class="scroll-bar-vertical-thumb"
      style="transform: translateY(0px); height: 0px; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
    />
  </div>
</div>
`;
